player/lua/osc.lua

```diff
@@ -393,6 +394,26 @@ function get_track(type)
     return 0
 end
 
+-- WindowControl helpers
+function window_controls_enabled()
+    val = user_opts.windowcontrols
+    if val == "auto" then
+        return mp.get_property("options/border") == "no"
+    elseif val == "yes" or val == "no" then
+        return val ~= "no"
+    else
+        msg.error("Invalid setting \""..val.."\" for windowcontrols")
+    end
+end
+
+function window_controls_alignment()
```

Comment:
I think validation of user options and runtime tests (`window_controls_enabled()`, `window_controls_alignment()`) should not be mixed. I think it would be better to validate/process user options only once after `read_options()`.

If `window_controls_enabled()` is ever called more than once (is it?), then the error could show more than once with the current code, which is unexpected IMO. Same for alignment.